File: Frontend/ICP/Point2PointICP.cpp
#include <pcl/io/pcd_io.h>
#include <pcl/point_types.h>
#include <pcl/registration/icp.h>
#include <pcl/filters/voxel_grid.h>
#include <pcl/visualization/pcl_visualizer.h>

int main(int argc, char** argv) {
    // Load the point cloud from a .bin file
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
    // Load your point cloud data here (you may need to convert .bin to .pcd)
    pcl::io::loadPCDFile("../../../../KITTI/lidar/pcd/000004.pcd", *cloud);

    // First downsampling
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_down1(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::VoxelGrid<pcl::PointXYZ> vg1;
    vg1.setInputCloud(cloud);
    vg1.setLeafSize(0.3f, 0.3f, 0.3f);  // Increased from 0.1
    vg1.filter(*cloud_down1);

    // Second downsampling
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_down2(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::VoxelGrid<pcl::PointXYZ> vg2;
    vg2.setInputCloud(cloud_down1);
    vg2.setLeafSize(0.5f, 0.5f, 0.5f);  // Increased from 0.2
    vg2.filter(*cloud_down2);

    // Apply a transformation to create the destination point cloud
    Eigen::Matrix4f transformation = Eigen::Matrix4f::Identity();
    transformation(0, 0) = 0.862; transformation(0, 1) = 0.011; transformation(0, 2) = -0.507; transformation(0, 3) = 0.5;
    transformation(1, 0) = -0.139; transformation(1, 1) = 0.967; transformation(1, 2) = -0.215; transformation(1, 3) = 0.7;
    transformation(2, 0) = 0.487; transformation(2, 1) = 0.255; transformation(2, 2) = 0.835; transformation(2, 3) = -1.4;

    // Transform the original point cloud
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_transformed(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::transformPointCloud(*cloud_down2, *cloud_transformed, transformation);

    // Perform point-to-point ICP
    pcl::IterativeClosestPoint<pcl::PointXYZ, pcl::PointXYZ> icp;
    icp.setMaximumIterations(100);  // Add maximum iteration parameter
    icp.setInputSource(cloud_down2);
    icp.setInputTarget(cloud_transformed);
    pcl::PointCloud<pcl::PointXYZ> Final;
    icp.align(Final);

    // Output the results
    std::cout << "ICP has converged:" << icp.hasConverged() << " score: " << icp.getFitnessScore() << std::endl;
    std::cout << "ICP transformation:\n" << icp.getFinalTransformation() << std::endl;

    // Check if point clouds are valid before visualization
    if (cloud_down2->empty() || cloud_transformed->empty() || Final.empty()) {
        std::cerr << "One of the point clouds is empty!" << std::endl;
        return -1;  // Exit if any point cloud is empty
    }

    // Visualize the point clouds
    pcl::visualization::PCLVisualizer viewer("3D Viewer");
    viewer.setBackgroundColor(0, 0, 0);  // Black background
    viewer.addPointCloud<pcl::PointXYZ>(cloud_down2, "source cloud");
    viewer.addPointCloud<pcl::PointXYZ>(cloud_transformed, "target cloud");
    viewer.addPointCloud<pcl::PointXYZ>(Final.makeShared(), "aligned cloud");
    viewer.setPointCloudRenderingProperties(pcl::visualization::PCL_VISUALIZER_POINT_SIZE, 2, "source cloud");
    viewer.setPointCloudRenderingProperties(pcl::visualization::PCL_VISUALIZER_POINT_SIZE, 2, "target cloud");
    viewer.setPointCloudRenderingProperties(pcl::visualization::PCL_VISUALIZER_POINT_SIZE, 2, "aligned cloud");

    // Spin the viewer until it is closed
    while (!viewer.wasStopped()) {
        viewer.spinOnce(100);
    }

    return 0;
}

File: Frontend/ICP/Point2PointICP.py
import numpy as np
import open3d as o3d
import copy

# Load the point cloud from the .bin file
points = np.fromfile('../../../KITTI/lidar/bin/000004.bin', dtype=np.float32).reshape(-1, 4)
pcd = o3d.geometry.PointCloud()
pcd.points = o3d.utility.Vector3dVector(points[:, :3])

# First downsampling
voxel_size1 = 0.3  # Increased from 0.1
pcd_down1 = pcd.voxel_down_sample(voxel_size=voxel_size1)

# Second downsampling
voxel_size2 = 0.5  # Increased from 0.2
pcd_down2 = pcd_down1.voxel_down_sample(voxel_size=voxel_size2)

# Apply a transformation to create the destination point cloud
transformation = np.array([
    [0.862, 0.011, -0.507, 0.5],
    [-0.139, 0.967, -0.215, 0.7],
    [0.487, 0.255, 0.835, -1.4],
    [0.0,   0.0,   0.0,    1.0]
])

# Copy and transform the original point cloud
pcd_transformed = copy.deepcopy(pcd_down2)
pcd_transformed.transform(transformation)

# Perform point-to-point ICP
trans_init = np.identity(4)  # Initial transformation guess (identity matrix)
threshold = 0.5  # Increased threshold for correspondence pairs
max_iteration = 100  # Add maximum iteration parameter
reg_p2p = o3d.pipelines.registration.registration_icp(
    pcd_down2, pcd_transformed,
    threshold, trans_init,
    o3d.pipelines.registration.TransformationEstimationPointToPoint(),
    o3d.pipelines.registration.ICPConvergenceCriteria(max_iteration=max_iteration))

print(reg_p2p)
print(f"Ground truth transformation:\n{transformation}")
print(f"ICP transformation:\n{reg_p2p.transformation}")

# Apply the ICP result to align the source to target
pcd_aligned = copy.deepcopy(pcd_down2)
pcd_aligned.transform(reg_p2p.transformation)

# Paint the point clouds with different colors
pcd_down2.paint_uniform_color([1, 0, 0])          # Source: Red
pcd_transformed.paint_uniform_color([0, 1, 0])     # Target: Green
pcd_aligned.paint_uniform_color([0, 0, 1])         # Aligned Result: Blue

# Visualize all point clouds
vis = o3d.visualization.Visualizer()
vis.create_window()

# Add the geometries
vis.add_geometry(pcd_down2)
vis.add_geometry(pcd_transformed)
vis.add_geometry(pcd_aligned)

# Improve visualization settings
opt = vis.get_render_option()
opt.background_color = np.asarray([0, 0, 0])  # Black background
opt.point_size = 2.0  # Larger points

vis.run()
vis.destroy_window()
